### PR TITLE
Prevent perpetual conflicts in generated manifest files

### PR DESCRIPTION
**Pull request by levb**
No ticket - this is more of an open discussion. 

Ever since we included hashes in the manifest files, we get constant merge conflicts there. 
- We don't want to remove the source file from the repository - it would affect source analysis tools, "naked" `go test`, etc. 
- .gitignore does not appear to ignore them if they are already in the repo

We should all ignore further changes to these 2 files, and I only found a way to do it locally:
```
git update-index --assume-unchanged server/manifest.go
git update-index --assume-unchanged webapp/src/manifest.js
```

If anyone knows how to do it centrally on the repository itself, and effectively .gitignore without removing - please comment. @mickmister @jfrerich @crspeller

### CODE REVIEW

**Comment by cpoile** [comment]
If the files are ignored, we're not uploading changes made to them, right? What happens when we commit a new change and then package and release that version? The packaged plugin won't have the updated manifest file, and jira info won't give us what we need in the wild?
Or when the CI builds and packages the plugin, the manifest files will get updated?  
I guess I'm a little fuzzy about how the CI and packaging works. :)

**Comment by hanzei** [comment]
Let my try to understand the general idea here. I see that it's usefull to get the current hash. But why are you checking it in and including it in `manifest.go`? Why not just get the hash on build time?

**Comment by levb** [comment]
The ^^ `git` commands don't really seem todo much for me, my local changes still show up dirty.

I think it's ok to have the committed files with have empty strings or a fixed value, since the only real way to build is to `make`, and that updates the constants anyway. We can't just `git rm` the files and `.gitignore` them locally, and I have not found a better way yet. Will look some more.

I'd like to keep this PR open for discussion, until we find a solution. Having to re-checkout the files prior to each commit - which is what I started doing to keep them clean of changes/conflicts is effectively accomplishing the same, but laboriously.

**Comment by levb** [comment]
@hanzei we are generating all values every time `make` runs. The issue is, is we remove these files from the repo and rely on running `make` to generate them, then the committed source tree will not be compile-able on its own. So we should have _something_ committed. Then, how do you ignore all of your local changes to a committed file? `.gitignore` does not appear to apply in this case.

**Comment by hanzei** [comment]
Sorry for beeing unrelenting here, but why is `Hash` part of `manifest` and not just a global variable set on complie time, like in the server?

**Comment by hanzei** [comment]
I get the problem you are trying to fix with this PR. I'm just wondering if we can avoide the problem altogether.

**Comment by levb** [comment]
You're right, that'd be a better way, I'll change to that, unless @cpoile has other considerations.

**Comment by levb** [comment]
@cpoile @hanzei ready for re-review

**Comment by mickmister** [comment]
@cpoile I believe there will always be the correct hash available when the plugin is running. When we compile for release, the hash will automatically be created and included.

The only time the hash would be wrong is when we're running tests, where we have not compiled the correct hash beforehand.

File: Makefile [hunk 1]
```diff
@@ -287,3 +304,15 @@ kill-plugin:
 		echo "Killing plugin pid $$PID"; \
 		kill -9 $$PID; \
 	done; \
+
+.PHONY: validate-go-version
+validate-go-version: ## Validates the installed version of go against Mattermost's minimum requirement.
```
Comment (hanzei) [inline Makefile]: Why is this needed for this PR? I'm not against the change in general, but IMO it shouldn't be part of this PR.

File: Makefile [hunk 2]
```diff
@@ -1,10 +1,27 @@
...
+LDFLAGS += -X "main.BuildHash=$(BUILD_HASH)"
+LDFLAGS += -X "main.BuildHashShort=$(BUILD_HASH_SHORT)"
+GO_MAJOR_VERSION = $(shell $(GO) version | cut -c 14- | cut -d' ' -f1 | cut -d'.' -f1)
+GO_MINOR_VERSION = $(shell $(GO) version | cut -c 14- | cut -d' ' -f1 | cut -d'.' -f2)
+GO_VERSION_VALIDATION_ERR_MSG = Golang version is not supported, please update to at least $(MINIMUM_SUPPORTED_GO_MAJOR_VERSION).$(MINIMUM_SUPPORTED_GO_MINOR_VERSION)
+
+GOVET = $(GO) vet $(GOFLAGS)
+GOTOOL = $(GO) tool $(GOFLAGS)
+GOBUILD = $(GO) build $(GOFLAGS) -ldflags '$(LDFLAGS)'
+GOTEST = $(GO) test $(GOFLAGS) -ldflags '$(LDFLAGS)'
```
Comment (hanzei) [inline Makefile]: Would you please submit these changes to https://github.com/mattermost/mattermost-plugin-starter-template? I would like to discuss them desperately.

File: Makefile [hunk 3]
```diff
@@ -1,10 +1,27 @@
...
 MM_UTILITIES_DIR ?= ../mattermost-utilities
 
+MINIMUM_SUPPORTED_GO_MAJOR_VERSION = 1
+MINIMUM_SUPPORTED_GO_MINOR_VERSION = 12
 export GO111MODULE=on
+GO ?= $(shell command -v go 2> /dev/null)
+GOFLAGS ?= $(GOFLAGS:)
+BUILD_DATE = $(shell date -u)
+BUILD_HASH = $(shell git rev-parse HEAD)
+BUILD_HASH_SHORT = $(shell git rev-parse --short HEAD)
```
Comment (hanzei) [inline Makefile]: :+1:

File: Makefile (hunk 1 above)
Comment (levb) [reply Makefile]: I was lifting worthy go pieces from mattermost-server, this one came with it, why not?

I'll make a separate PR for the template, but I'd like this one to merge very soon since the merge conflicts are really annoying.

File: Makefile [hunk 4]
```diff
@@ -1,10 +1,27 @@
...
+LDFLAGS += -X "main.BuildDate=$(BUILD_DATE)"
+LDFLAGS += -X "main.BuildHash=$(BUILD_HASH)"
+LDFLAGS += -X "main.BuildHashShort=$(BUILD_HASH_SHORT)"
```
Comment (cpoile) [inline Makefile]: Ah, yah, this is much better. Thank you!
I should have thought of that... <facepalm>

**Comment by cpoile** [comment]
@hanzei I can't tell if you've approved or not. If you have, can you merge? Thanks!

File: Makefile (hunk 1 above)
Comment (hanzei) [reply Makefile]: Let's merge this one now and discuss these changes in depth on the PR for the template repo.

**Comment by cpoile** [comment]
@levb hah... ahh... We need to solve the merge conflict. I'll try to do that manually here.

**Comment by hanzei** [comment]
I'm working on it

**Comment by levb** [comment]
LOL, and I was about to jump in and fix the merge conflict! :)

**Comment by hanzei** [comment]
We are **really** hard to get this merged soon :laughing:

**Comment by cpoile** [comment]
:)
